rfd/0192-oci-join.md

```diff
@@ -0,0 +1,252 @@
+
+### Implementation
+
+#### Token spec
+
+The provision token will be extended to include a new `oracle` section:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example-oci-token
+spec:
+  roles: [Node, Kube, Db]
+  oracle:
+    allow:
+        # OCID of the tenancy to allow instances to join from. Required.
+      - tenancy: "ocid1.tenancy.oc1..<unique ID>"
+        # Compartments to allow instances to join from. Only the direct parent
+        # compartment applies; i.e. nested compartments are not taken into account.
+        # May be specified by name or by OCID. If compartments is empty or a wildcard,
+        # instances can join from any compartment in the tenancy.
+        parent_compartments: ["my_compartment", "ocid1.compartment.oc1...<unique_ID>"]
+        # Regions to allow instances to join from. Both full names ("us-phoenix-1")
+        # and abbreviations ("phx") are allowed. If regions is empty or
+        # a wildcard, instances can join from any region.
+        regions: ["phx", "us-ashburn-1"]
+        # Add more entries as necessary.
+      - tenancy: "..."
+        parent_compartments: ["foo", "bar"]
+        regions: ["baz", "quux"]
+        # ...
+```
+
+#### Join process
+
+When a node initiates the Oracle join method:
+
+- The node first fetches credentials for its
+[instance principal](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm#concepts)
+via the Oracle instance metadata service.
+- With those credentials, the node will create a
+[signed HTTP request](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/signingrequests.htm)
+to the Oracle Cloud API to
+[fetch the compartment the instance is in](https://docs.oracle.com/en-us/iaas/api/#/en/identity/20160918/Compartment/GetCompartment),
+at `https://iaas.{region}.oraclecloud.com/{api version}/compartment/{compartment's OCID}`.
+The instance's principal does not need any additional permissions to make this request.
+- The node will then make a `RegisterUsingToken` request to the auth server and
+sends the URL and headers of the signed request as parameters (the auth server
+will be able to recreate the rest of the request).
+
+When the auth server receives a `RegisterUsingToken` request for the Oracle join method:
+
+- The auth server extracts the key ID from the provided Authorization header. The key
+ID is a string that Oracle uses to identify the caller. For instance principals,
+the key ID is a JWT prefixed by the string `ST$` (unfortunately I could not
+find docs that back this up, I found this experimentally).
+- The auth server decodes the JWT and maps the claims `opc-tenant`,
+`opc-compartment`, and `opc-instance` to the instance's tenancy ID, compartment
+ID, and instance ID respectively.
+- The auth server verifies that the compartment ID and region in the provided
+API URL match the compartment ID and region from the claims (the region can be
+extracted from the instance ID).
```

Comment:
I'm not finding much/anything about this JWT in the request signing docs. If it's not well documented and seems like an implementation detail, I'm wondering what benefit we get from parsing it and comparing it with the URL? Does it give us anything in addition to what we would get by just ignoring the JWT?

Reply:
If we didn't look at it, we wouldn't know which instance we were talking to (or its compartment, tenancy, etc.). All we would know is that someone is allowed to list compartments.